### DISCUSSION TITLE
multi UEs+ZeroMQ+multi gnbs +open5gs

### DISCUSSION THREAD

**Discussion by wmwwimor** [srsran/srsRAN_Project#131] [General Help] [answered]
I would like to know if it is possible to implement this architecture:multi UEs+ZeroMQ+one gnb +open5gs. Can it currently be achieved using GNU Radio? Is it possible to further expand it into such an architecture:multi UEs+ZeroMQ+multi gnbs +open5gs? I don't want to use physical devices. Is there something similar to ZeroMQ available now that can assist in simulating radio frequencies?

**Comment by brendan-mcauliffe** [comment, marked as answer]
See answer to the same question here: https://github.com/srsran/srsRAN_4G/discussions/1203#discussioncomment-6465385

**Comment by jonniebeeb** [comment]
Can I ask a follow up question here please.

I have tried implementing the above suggestion, using a GNURadio flowgraph to connect two UEs. I am running a 5G SA gNB using the srsRAN_Project. I can connect a single UE to the gNB, using ZeroMQ. It connects, my UE gets an IP address and I can ping the UE from the UPF and vice versa so my basic setyup is working.

I have configured a second UE, setup a second namespace, allocated a second IMSI etc. 

When I start the flowgraph, the UEs attach and I get RRC connected. However neither gets an IP address - I don't see "PDU Session Establishment successful". The issue seems to be that each UE gets the same ra-rnti and c-rnti number as below :

UE1
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0xf, tti=171
Random Access Complete.     c-rnti=0x4601, ta=32
RRC Connected

UE2
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0xf, tti=171
Random Access Complete.     c-rnti=0x4601, ta=32
RRC Connected

I think this is because each UE is connecting at the same time to the gNB. If I add a delay block into one of the UE paths in the flowgraph, I can get one UE to successfully get an IP address, although both UEs still get the same ra-rnti and c-rnti numbers.

UE1
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected

UE2
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.3
RRC NR reconfiguration successful.


Is there something else I need to configure? My understanding was that if two UEs try to connect to the gNB at the same time, they should each implement a random back off and try again.

Thanks

**Reply by wmwwimor** [reply]
I have the same issue as yours.Have you solved it?

**Reply by jonniebeeb** [reply]
Not yet. I'm going to look at the PRACH process. I think this is the cause of the issue. I think it's because both UEs do the same thing at exactly the same time.

**Reply by wmwwimor** [reply]
I also add a delay block to the flow graph, but both ues still can't be connected to the 5GC ,both of them stuck in "RRC connected".
And I got the log from open5gs :07/31 16:08:32.075: [amf] INFO: InitialUEMessage (../src/amf/ngap-handler.c:372)
07/31 16:08:32.075: [amf] INFO: [Added] Number of gNB-UEs is now 1 (../src/amf/context.c:2510)
07/31 16:08:32.075: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] TAC[7] CellID[0x0] (../src/amf/ngap-handler.c:533)
07/31 16:08:32.076: [amf] INFO: [suci-0-001-01-0000-0-0-0123456780] Unknown UE by SUCI (../src/amf/context.c:1776)
07/31 16:08:32.076: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1563)
07/31 16:08:32.076: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:985)
07/31 16:08:32.076: [gmm] INFO: [suci-0-001-01-0000-0-0-0123456780]    SUCI (../src/amf/gmm-handler.c:152)
07/31 16:08:32.140: [gmm] WARNING: Authentication failure(MAC failure) (../src/amf/gmm-sm.c:1348)
07/31 16:08:32.140: [amf] WARNING: [suci-0-001-01-0000-0-0-0123456780] Authentication reject (../src/amf/nas-path.c:538)
07/31 16:08:32.141: [amf] INFO: UE Context Release [Action:3] (../src/amf/ngap-handler.c:1632)
07/31 16:08:32.141: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] (../src/amf/ngap-handler.c:1633)
07/31 16:08:32.141: [amf] INFO:     SUCI[suci-0-001-01-0000-0-0-0123456780] (../src/amf/ngap-handler.c:1636)
07/31 16:08:32.141: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2517)
07/31 16:08:32.141: [amf] INFO: [Removed] Number of AMF-UEs is now 0 (../src/amf/context.c:1656)

It seems that UE1 has been connected ,but is also discarded.
Which part of the path you place the delay block? 
Thanks!!!

**Reply by jonniebeeb** [reply]
This is what I've tried, see if it helps for you.

![two_ue1024_1](https://github.com/srsran/srsRAN_Project/assets/75319802/e93fc55f-5035-4d93-be71-32ae47618e66)

**Reply by wmwwimor** [reply]
Thanks for your help! One of UEs  got an IP .
I noticed that the warning "Authentication failure(MAC failure) (../src/amf/gmm-sm.c:1348)" of open5gs may be a hint to solve the multiple ues problems.  In my opinion, the connetions between gnb and ue ,gnb and 5gc are ok ,we should revise the part of open5gs and ue?

**Reply by wmwwimor** [reply]
I made a test:open5gs always runs in the background:

1. Start gNB and UE1 separately in the terminal.

2. Connect gNB and UE1 in GNURadio.

UE1 gets IP address 10.45.0.2.

3. Disconnect the connection between gNB and UE1. Close UE1 first, then close gNB.

4. Restart gNB and then start UE2.

5. Connect gNB and UE2 in GNURadio.

UE2 gets IP address 10.45.0.3.

The log from open5gs :                                                                             08/03 15:12:52.782: [amf] INFO: [Added] Number of AMF-Sessions is now 2 (../src/amf/context.c:2531)
08/03 15:12:52.782: [gmm] INFO: UE SUPI[imsi-001010123456781] DNN[srsapn] S_NSSAI[SST:1 SD:0xffffff] (../src/amf/gmm-handler.c:1236)
08/03 15:12:52.783: [smf] INFO: [Added] Number of SMF-UEs is now 2 (../src/smf/context.c:1010)
08/03 15:12:52.783: [smf] INFO: [Added] Number of SMF-Sessions is now 2 (../src/smf/context.c:3050)
08/03 15:12:52.787: [smf] INFO: UE SUPI[imsi-001010123456781] DNN[srsapn] IPv4[10.45.0.3] IPv6[] (../src/smf/npcf-handler.c:539)
08/03 15:12:52.788: [upf] INFO: [Added] Number of UPF-Sessions is now 2 (../src/upf/context.c:206)                                                                                           From srsran website:# **Known issues**

- For a clean tear down, the UE needs to be terminated first, then the gNB.
- gNB and UE can only run once, after the UE has been detached, the gNB needs to be restarted.



I found that even after releasing UE1 and restarting gNB, open5gs still retains the context of UE1.

**Reply by jonniebeeb** [reply]
Hi @wmwwimor 

I think I have managed to get two UEs to connect ! I think my idea that the UE's are sending the PRACH at exactly the same time with the same sequence is what is causing the problem.

Here is what I have done, see if you get the same as me. 

Firstly I have removed the delay I added into the Gnu flowgraph. I set the delays to 0ms. 

I have copied the srsue binary and called it srsue1. I will use this when starting ue1 with ue1.conf

I have then changed this code on the UE side (in srsRAN_4G). 

In prach.cc, in this function, I have tried a quick and dirty hack -

```
bool prach::is_ready_to_send(uint32_t current_tti_, uint32_t current_pci)
{
  static int hold_off = 0;
  // Make sure the curernt PCI is the one we configured the PRACH for
  if (is_pending() && current_pci == cell.id) {

    if (hold_off < 100 )
      hold_off++;

    std::lock_guard<std::mutex> lock(mutex);
    // consider the number of subframes the transmission must be anticipated
    if(hold_off > 50){
      uint32_t tti_tx = TTI_TX(current_tti_);
      if (srsran_prach_tti_opportunity(&prach_obj, tti_tx, allowed_subframe)) {
        Debug("PRACH Buffer: Ready to send at tti: %d (now is %d)", tti_tx, current_tti_);
        transmitted_tti = tti_tx;
        return true;
      }
    }
  }
  return false;
}
```

This binary I rename srsue2 and I use this when starting the second ue with ue2.conf

I have basically added a hold off so that the second ue waits before starting the PRACH process.


I know get the two UEs to connect and I get an IP address for each one ! Each gets a different c-rnti number, although they both get the same ra-rnti number - not sure if that is still an issue.

UE1 

Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.4
RRC NR reconfiguration successful.

UE2 

Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=494
Random Access Complete.     c-rnti=0x4602, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.3
RRC NR reconfiguration successful.

**Reply by jonniebeeb** [reply]
Quick update, I have extended this to three UEs now, using a different hold_off for the third UE, so I think my approach is correct. It needs tidying up, so the hold_off can be passed in as a parameter

```
bool prach::is_ready_to_send(uint32_t current_tti_, uint32_t current_pci)
{
  static int hold_off = 0;
  // Make sure the curernt PCI is the one we configured the PRACH for
  if (is_pending() && current_pci == cell.id) {

    if (hold_off < 400 )
      hold_off++;

    std::lock_guard<std::mutex> lock(mutex);
    // consider the number of subframes the transmission must be anticipated
    if(hold_off > 200){
      uint32_t tti_tx = TTI_TX(current_tti_);
      if (srsran_prach_tti_opportunity(&prach_obj, tti_tx, allowed_subframe)) {
        Debug("PRACH Buffer: Ready to send at tti: %d (now is %d)", tti_tx, current_tti_);
        transmitted_tti = tti_tx;
        return true;
      }
    }
  }
  return false;
}
```

UE1
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.3
RRC NR reconfiguration successful.

UE2
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=494
Random Access Complete.     c-rnti=0x4602, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.4
RRC NR reconfiguration successful.

UE3
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=654
Random Access Complete.     c-rnti=0x4603, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.5
RRC NR reconfiguration successful.

**Reply by wmwwimor** [reply]
Thanks for your reply!!
I have a try .But it seems that I didn't solve the RRC collision.
As what you said ,I revise the prach.cc, then remake the srsRAN_4G, so I get a new binary file :srsue, and rename it as srsue2,the orginal one I name it as srsue1 .
When I start up 2 ues in different terminal ,and run the gnuradio removed the delay. block ,I still find that both of them stuck at "RRC connected"，the "tti" and "c-rnti" are the same.
So I'd like to know if my operation is right？
Thank you!!!

**Reply by jonniebeeb** [reply]
what tti do you get? The change in prach.cc should force the second UE to wait so it should have a different tti.

Be sure to copy the original srsue binary to srsue1 before making any code changes.

**Reply by jonniebeeb** [reply]
and has the tti changed from what you had before the changes?

**Reply by wmwwimor** [reply]
After I revise  the original prach .cc ,I rebuild and remake the srsRAN_4G in ubuntu, so I can get new binary file srsue,renamed as srsue2, used for running ue2,

The Open5gs \  srsRAN_Project \ srsRAN_4G(The original one as UE1) \srsRAN_4G(as UE2) run in different teminal , and then I start up the gnuradio.

Following these orders in different terminal:
~/open5gs$ sudo ./net5g.sh 
~/srsRAN_Project/build/apps/gnb$ sudo ./gnb -c gnb.yml
~/srsRAN_4G/build/srsue/src$ sudo ./srsue1 ./ue1.conf
~/Downloads/srsRAN_4G/build/srsue/src$ sudo ./srsue2 ./ue2conf

This is what I get from 2ues:
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
_______________________________________________________________________________________________________________
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected

I don't know why I still have the same tti from 2UEs like before.

**Reply by wmwwimor** [reply]
Have you ever saved the original prach.cc used for ue1,and then rebuild the srsRAN_4G to get new one used for ue2,because you change the  source code.
Waiting for your reply!!!

**Reply by jonniebeeb** [reply]
I save the srsue binary and name it srsue1 before changing anything. Then I change prach.cc, rebuild and the new srsue binary I call srsue2. You need to run two different binaries to make this work.

**Reply by jonniebeeb** [reply]
Not sure why you are seeing a tti=334 on both UEs. If the change in prach.cc is successful, it should cause the tti to increase. It's as if the changes you've made aren't being used. I'd say check when you rebuild srsRAN_4G that it is rebuilding the srsue binary.

**Reply by wmwwimor** [reply]
I check the cmakelist in srsRAN_4G，and I'm sure the prach.cc is included in makefile. Besides ,I check the log printed when I remake the  srsRAN_4G ，it also include  the file prach.cc.
So I think there is sth wrong when I rebuild the project.
I removed the old binary files ,and followed your guide again, I got the same result as yours. Thanks so much !!!
You are right . After sorting out this problem, maybe more gnbs can be expanded.
Thanks for your help again!!

**Reply by jonniebeeb** [reply]
I’m puzzled as to why it’s not working for you. I have repeated this on a virtual machine I have in the cloud and it works for me, although I have another issue here in that UE3 failed to establish a PDU session which I can see in the amf logs. But the UE does get a unique c-rnti number. Can you put a printf statement in the prach.cc to check what the hold_off count is doing?

**Reply by jonniebeeb** [reply]
In prach.cc have you included the conditional statement 
If(hold_off > 50)

**Comment by pgawlowicz** [comment]
Could you share configs that you use for each UE?

You can also have a look at this tutorial:
https://openaicellular.github.io/oaic/multi_ue_example.html

It uses the srsenb (4G), but it should be possible to connect the new gnb (5G) using the same approach.

**Reply by pgawlowicz** [reply]
@wmwwimor @jonniebeeb Please check the new section of our gnb+srsUE tutorial that describes how to emulate multiple-UEs:
https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation

Please remember to clone the newest version of the srsRAN-4G repository.
